flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/Regularization.scala

```diff
@@ -0,0 +1,207 @@
+    * @return The loss value with regularization applied.
+    */
+  override def regLoss(loss: Double, weightVector: FlinkVector, regParameter: Double)
+    : Double = {
+    val brzVector = weightVector.asBreeze
+    loss + regParameter * (brzVector dot brzVector) / 2
+  }
+
+  /** Adds the regularization gradient term to the loss gradient. The gradient is updated in place.
+    *
+    * @param weightVector The current vector of weights.
+    * @param lossGradient The loss gradient, without regularization. Updated in-place.
+    * @param regParameter The regularization parameter, $\lambda$.
+    */
+  override def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regParameter: Double): Unit = {
+    BLAS.axpy(regParameter, weightVector, lossGradient)
+  }
+}
+
+/** $L_1$ regularization penalty.
+  *
+  * The $L_1$ penalty can be used to drive a number of the solution coefficients to 0, thereby
+  * producing sparse solutions.
+  *
+  */
+class L1Regularization extends Regularization {
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * Implementation was taken from the Apache Spark Mllib library:
+    * http://git.io/vfZIT
+    *
+    * @param oldWeights The weights to be updated
+    * @param gradient The gradient according to which we will update the weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The regularization parameter to be applied in the case of L1
+    *                     regularization
+    */
+  override def takeStep(
+      oldWeights: FlinkVector,
+      gradient: FlinkVector,
+      effectiveStepSize: Double,
+      regParameter: Double) {
+    BLAS.axpy(-effectiveStepSize, gradient, oldWeights)
+    val brzWeights = oldWeights.asBreeze
+
+    // Apply proximal operator (soft thresholding)
+    val shrinkageVal = regParameter * effectiveStepSize
+    var i = 0
+    while (i < brzWeights.length) {
+      val wi = brzWeights(i)
+      brzWeights(i) = signum(wi) * math.max(0.0, abs(wi) - shrinkageVal)
+      i += 1
+    }
+
+    BLAS.copy(brzWeights.fromBreeze, oldWeights)
+
+    // We could maybe define a Breeze Universal function for the proximal operator, and test if it's
+    // faster that the for loop + copy above
+    //    brzWeights = signum(brzWeights) * max(0.0, abs(brzWeights) - shrinkageVal)
+
+  }
+
+  /** Adds the regularization term to the loss value
+    *
+    * @param loss The loss value, before applying regularization.
+    * @param weightVector The current vector of weights.
+    * @param regularizationParameter The regularization parameter, $\lambda$.
+    * @return The loss value with regularization applied.
+    */
+  override def regLoss(loss: Double, weightVector: FlinkVector, regularizationParameter: Double):
+  Double = {
+    loss + BreezeNorm(weightVector.asBreeze, 1.0) * regularizationParameter
```

Comment:
For the 1-norm it doesn't use pow as far as I can tell:

https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/linalg/operators/DenseVectorOps.scala#L702
Snippet:
```scala
  implicit def canNormField[T:Field]: norm.Impl2[DenseVector[T],Double,Double] = {
    val f = implicitly[Field[T]]
    new norm.Impl2[DenseVector[T],Double,Double] {
      def apply(v: DenseVector[T],n: Double) = {
        import v._
        if (n == 1) {
          var sum = 0.0
          foreach (v => sum += f.sNorm(v) )
          sum
        } else if (n == 2) {
          var sum = 0.0
          foreach (v => { val nn = f.sNorm(v); sum += nn * nn })
          math.sqrt(sum)
        } else if (n == Double.PositiveInfinity) {
          var max = 0.0
          foreach (v => { val nn = f.sNorm(v); if (nn > max) max = nn })
          max
        } else {
          var sum = 0.0
          foreach (v => { val nn = f.sNorm(v); sum += math.pow(nn,n) })
          math.pow(sum, 1.0 / n)
        }
      }
    }
  }
```